packages/cardpay-sdk/sdk/utils/module-utils.ts

```diff
@@ -0,0 +1,89 @@
+import { AbiItem } from 'web3-utils';
+import { getAddress } from '../../contracts/addresses';
+import ModuleProxyFactoryABI from '../../contracts/abi/modules/module-proxy-factory';
+import { Contract } from 'web3-eth-contract';
+import { ethers } from 'ethers';
+import { Operation } from './safe-utils';
+import multiSend from '../../contracts/abi/modules/multi-send';
+import { pack } from '@ethersproject/solidity';
+import { hexDataLength } from 'ethers/lib/utils';
+
+export interface Transaction {
+  to: string;
+  value: string;
+  data: string;
+  operation: Operation;
+}
+
+export async function deployAndSetUpModule(
+  web3: Web3,
+  masterCopy: Contract,
+  setupArgs: {
+    types: string[];
+    values: any[];
+  }
+) {
+  let factory = new web3.eth.Contract(ModuleProxyFactoryABI as AbiItem[], await getAddress('moduleProxyFactory', web3));
+
+  let encodeInitParams = web3.eth.abi.encodeParameters(setupArgs.types, setupArgs.values);
+  let moduleSetupData = masterCopy.methods.setUp(encodeInitParams).encodeABI();
+  let saltNonce = new Date().getTime().toString();
+  let expectedModuleAddress = await calculateProxyAddress(
+    factory,
+    masterCopy.options.address,
+    moduleSetupData,
+    saltNonce
+  );
+
+  let deployData = factory.methods.deployModule(masterCopy.options.address, moduleSetupData, saltNonce).encodeABI();
+  let transaction: Transaction = {
+    data: deployData,
+    to: factory.options.address,
+    value: '0',
+    operation: Operation.CALL,
+  };
+  return {
+    transaction,
+    expectedModuleAddress,
+  };
+}
+
+export async function calculateProxyAddress(
+  factory: Contract,
+  masterCopyAddress: string,
+  initData: string,
+  saltNonce: string
+) {
+  masterCopyAddress = masterCopyAddress.toLowerCase().replace(/^0x/, '');
+  const byteCode = '0x602d8060093d393df3363d3d373d3d3d363d73' + masterCopyAddress + '5af43d82803e903d91602b57fd5bf3';
+
+  const salt = ethers.utils.solidityKeccak256(
+    ['bytes32', 'uint256'],
+    [ethers.utils.solidityKeccak256(['bytes'], [initData]), saltNonce]
+  );
+
+  return ethers.utils.getCreate2Address(factory.options.address, salt, ethers.utils.keccak256(byteCode));
+}
+
+export async function encodeMultiSend(web3: Web3, transactions: readonly Transaction[]): Promise<Transaction> {
+  const transactionsEncoded = '0x' + transactions.map(encodePacked).map(remove0x).join('');
+
+  const multiSendContract = new web3.eth.Contract(multiSend as AbiItem[], await getAddress('multiSend', web3));
+  const data = multiSendContract.methods.multiSend(transactionsEncoded).encodeABI();
+
+  return {
+    operation: Operation.DELEGATECALL,
+    to: multiSendContract.options.address,
+    value: '0',
+    data,
+  };
+}
+
+const encodePacked = (tx: any) =>
+  pack(
+    ['uint8', 'address', 'uint256', 'uint256', 'bytes'],
+    [tx.operation || Operation.CALL, tx.to, tx.value, hexDataLength(tx.data), tx.data]
+  );
+
+const remove0x = (hexString: string) => hexString.substr(2);
```

Comment:
for these 2 functions is it necessary to use `=>`? fat arrows perform a .`bind()` to the calling context, but we are not using `this` in either of the functions. Probably we can just use normal module scoped functions here:
```js
function encodePacked() { ... }
function remove0x() { ... }
```

Also for the type of `tx`, could we use a better type than `any`? In the `encodeMultiSend()` you are using a `Transaction[]` type--is this not the type that we should be using?.

Also, you couse use an idempotent remove0x() that use the logic `hexString.replace(/^0x/, '')` this logic could be run mulitple times against the hex and always return the current value, where `hexString.substr(2)` will not.

Reply:
Updated